File: ts/src/tools/get-case-comments.ts
import { z } from "zod";
import type { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
import type { FogBugzClient } from "../fogbugz-client.js";
import type { FogBugzEvent } from "../types.js";

const EVENT_TYPE_NAMES: Record<number, string> = {
  1: "Opened",
  2: "Edited",
  3: "Assigned",
  4: "Reactivated",
  5: "Reopened",
  6: "Closed",
  7: "Moved",
  8: "Unknown",
  9: "Replied",
  10: "Forwarded",
  11: "Received",
  12: "Sorted",
  13: "Not Sorted",
  14: "Resolved",
  15: "Emailed",
  16: "Release Noted",
  17: "Deleted Attachment",
};

function formatEvent(event: FogBugzEvent, baseUrl: string): string {
  const eventType = EVENT_TYPE_NAMES[event.evt] ?? `Event ${event.evt}`;
  const date = event.dt
    ? new Date(event.dt).toLocaleString("en-US", {
        dateStyle: "medium",
        timeStyle: "short",
      })
    : "—";

  const lines: string[] = [
    `### ${eventType} by ${event.sPerson} — ${date}`,
    event.sVerb ? `_${event.sVerb}_` : "",
  ];

  if (event.sChanges) {
    lines.push(`**Changes:** ${event.sChanges}`);
  }

  const text = event.s?.trim();
  if (text) {
    lines.push("", text);
  }

  if (event.rgAttachments?.length) {
    lines.push("", "**Attachments:**");
    for (const att of event.rgAttachments) {
      const cleanUrl = att.sURL.replace(/&amp;/g, "&");
      lines.push(`- ${att.sFileName}: ${baseUrl}/${cleanUrl}`);
    }
  }

  return lines.filter((l) => l !== "").join("\n");
}

export function registerGetCaseComments(
  server: McpServer,
  client: FogBugzClient,
) {
  server.tool(
    "get_case_comments",
    "Get all comments and events for a FogBugz case, including the full text of each comment, who made it, when, what changed, and any attachments.",
    {
      case_number: z.number().describe("The FogBugz case number"),
    },
    { readOnlyHint: true },
    async ({ case_number }) => {
      const data = await client.search(
        String(case_number),
        ["sTitle", "events"],
        1,
      );
      if (data.cases.length === 0) {
        return {
          content: [
            { type: "text", text: `Case ${case_number} not found.` },
          ],
        };
      }

      const c = data.cases[0];
      const caseUrl = client.getCaseUrl(c.ixBug);
      const baseUrl = caseUrl.replace(`/f/cases/${c.ixBug}`, "");

      const events = c.events ?? [];
      if (events.length === 0) {
        return {
          content: [
            {
              type: "text",
              text: `Case ${case_number} (${c.sTitle}) has no events.`,
            },
          ],
        };
      }

      const header = `# Comments for Case ${c.ixBug}: ${c.sTitle ?? ""}\nLink: ${caseUrl}\n${events.length} event(s)\n`;
      const body = events
        .map((e) => formatEvent(e, baseUrl))
        .join("\n\n---\n\n");

      return { content: [{ type: "text", text: header + "\n" + body }] };
    },
  );
}

File: python/src/fogbugz_mcp/tools/get_case_comments.py
from __future__ import annotations

from datetime import datetime
from typing import TYPE_CHECKING

from mcp.server.fastmcp import FastMCP

if TYPE_CHECKING:
    from ..fogbugz_client import FogBugzClient

EVENT_TYPE_NAMES: dict[int, str] = {
    1: "Opened",
    2: "Edited",
    3: "Assigned",
    4: "Reactivated",
    5: "Reopened",
    6: "Closed",
    7: "Moved",
    8: "Unknown",
    9: "Replied",
    10: "Forwarded",
    11: "Received",
    12: "Sorted",
    13: "Not Sorted",
    14: "Resolved",
    15: "Emailed",
    16: "Release Noted",
    17: "Deleted Attachment",
}


def _format_event(event: dict, base_url: str) -> str:
    event_type = EVENT_TYPE_NAMES.get(event.get("evt", 0), f"Event {event.get('evt', '?')}")
    dt = event.get("dt")
    if dt:
        try:
            date_str = datetime.fromisoformat(dt.replace("Z", "+00:00")).strftime(
                "%b %d, %Y, %I:%M %p"
            )
        except (ValueError, AttributeError):
            date_str = dt
    else:
        date_str = "—"

    lines: list[str] = [
        f"### {event_type} by {event.get('sPerson', '?')} — {date_str}",
    ]

    if event.get("sVerb"):
        lines.append(f"_{event['sVerb']}_")

    if event.get("sChanges"):
        lines.append(f"**Changes:** {event['sChanges']}")

    text = (event.get("s") or "").strip()
    if text:
        lines.append("")
        lines.append(text)

    attachments = event.get("rgAttachments") or []
    if attachments:
        lines.append("")
        lines.append("**Attachments:**")
        for att in attachments:
            clean_url = att["sURL"].replace("&amp;", "&")
            lines.append(f"- {att['sFileName']}: {base_url}/{clean_url}")

    return "\n".join(line for line in lines if line != "" or lines.index(line) == lines.index(line))


def register_get_case_comments(mcp: FastMCP, client: FogBugzClient) -> None:
    @mcp.tool(
        name="get_case_comments",
        description=(
            "Get all comments and events for a FogBugz case, including the full text "
            "of each comment, who made it, when, what changed, and any attachments."
        ),
        annotations={"readOnlyHint": True},
    )
    async def get_case_comments(case_number: int) -> str:
        """Get all comments and events for a FogBugz case."""
        data = await client.search(str(case_number), ["sTitle", "events"], 1)
        cases = data.get("cases") or []
        if not cases:
            return f"Case {case_number} not found."

        c = cases[0]
        case_url = client.get_case_url(c["ixBug"])
        base_url = case_url.replace(f'/f/cases/{c["ixBug"]}', "")

        events = c.get("events") or []
        if not events:
            return f"Case {case_number} ({c.get('sTitle', '')}) has no events."

        header = (
            f"# Comments for Case {c['ixBug']}: {c.get('sTitle', '')}\n"
            f"Link: {case_url}\n"
            f"{len(events)} event(s)\n"
        )
        body = "\n\n---\n\n".join(_format_event(e, base_url) for e in events)
        return header + "\n" + body
